server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java

```diff
@@ -280,61 +278,68 @@ public void run() {
         }
     }
 
-    public void runOnApplierThread(final String source, Consumer<ClusterState> clusterStateConsumer,
-                                   final ClusterApplyListener listener, Priority priority) {
-        submitStateUpdateTask(source, ClusterStateTaskConfig.build(priority),
+    public void runOnApplierThread(
+        String source,
+        Priority priority,
```

Comment:
This is kind of a painful area. In fact we only need `runOnApplierThread` so that `IndicesStore` can delete shard data on this thread, which is needed because there's no other synchronization to prevent a shard from being allocated back to the node while it's being deleted. I'd rather fix that and then this whole thing just evaporates -- indeed once that has gone I think we might not even need a separate applier thread any more. But I'm not going to be doing that here. I added a comment to warn others against calling this anyway.